x-pack/plugins/ingest_manager/server/services/epm/archive/index.ts

```diff
@@ -0,0 +1,320 @@
+ */
+
+import yaml from 'js-yaml';
+import { uniq } from 'lodash';
+import { title } from 'process';
+
+import {
+  ArchivePackage,
+  Dataset,
+  RegistryConfigTemplate,
+  RegistryInput,
+  RegistryStream,
+  RegistryVarsEntry,
+} from '../../../../common/types';
+import { PackageInvalidArchiveError, PackageUnsupportedMediaTypeError } from '../../../errors';
+import { cacheGet, cacheSet, setArchiveFilelist } from '../registry/cache';
+import { unzipBuffer, untarBuffer, ArchiveEntry } from '../registry/extract';
+
+export async function loadArchivePackage({
+  archiveBuffer,
+  contentType,
+}: {
+  archiveBuffer: Buffer;
+  contentType: string;
+}): Promise<{ paths: string[]; archivePackageInfo: ArchivePackage }> {
+  const paths = await unpackArchiveToCache(archiveBuffer, contentType);
+  const archivePackageInfo = parseAndVerifyArchive(paths);
+  setArchiveFilelist(archivePackageInfo.name, archivePackageInfo.version, paths);
+
+  return {
+    paths,
+    archivePackageInfo,
+  };
+}
+
+function getBufferExtractorForContentType(contentType: string) {
+  if (contentType === 'application/gzip') {
+    return untarBuffer;
+  } else if (contentType === 'application/zip') {
+    return unzipBuffer;
+  } else {
+    throw new PackageUnsupportedMediaTypeError(
+      `Unsupported media type ${contentType}. Please use 'application/gzip' or 'application/zip'`
+    );
+  }
+}
+
+export async function unpackArchiveToCache(
+  archiveBuffer: Buffer,
+  contentType: string,
+  filter = (entry: ArchiveEntry): boolean => true
+): Promise<string[]> {
+  const bufferExtractor = getBufferExtractorForContentType(contentType);
+  const paths: string[] = [];
+  try {
+    await bufferExtractor(archiveBuffer, filter, (entry: ArchiveEntry) => {
+      const { path, buffer } = entry;
+      // skip directories
+      if (path.slice(-1) === '/') return;
+      if (buffer) {
+        cacheSet(path, buffer);
+        paths.push(path);
+      }
+    });
+  } catch (error) {
+    throw new PackageInvalidArchiveError(
+      `Error during extraction of uploaded package: ${error}. Assumed content type was ${contentType}, check if this matches the archive type.`
+    );
+  }
+
+  // While unpacking a tar.gz file with unzipBuffer() will result in a thrown error in the try-catch above,
+  // unpacking a zip file with untarBuffer() just results in nothing.
+  if (paths.length === 0) {
+    throw new PackageInvalidArchiveError(
+      `Uploaded archive seems empty. Assumed content type was ${contentType}, check if this matches the archive type.`
+    );
+  }
+  return paths;
+}
+
+// TODO: everything below performs verification of manifest.yml files, and hence duplicates functionality already implemented in the
+// package registry. At some point this should probably be replaced (or enhanced) with verification based on
+// https://github.com/elastic/package-spec/
+
+function parseAndVerifyArchive(paths: string[]): ArchivePackage {
```

Comment:
Nit: I wonder if we should put this in its own file called "validation" or similar to have the code around validation in its own place.

Reply:
Good point, there's a lot of `parseAndVerify` in that file now.